### ISSUE TITLE
ModelMatrix should be refactored

### ISSUE THREAD

**Issue by dmbates** [JuliaStats/DataFrames.jl#1012]
As I am responsible for the original, somewhat incomprehensible, version, I will create an update.

**Comment by GordStephen** [comment]
Probably worth coordinating this with the work in https://github.com/JuliaStats/DataFrames.jl/pull/870

**Comment by kleinschmidt** [comment]
I was thinking the same thing. I've nearly finished my dissertation so I'll have some free cycles. @dmbates perhaps we should talk about what you're thinking for the refactor?

**Comment by dmbates** [comment]
I can create a pull request relative to the current master to do this refactoring.  But I don't want to get divergent branches relative to the CatergoricalArrays and NullableArrays changes.  Any recommendations on how to coordinate with those changes and with the specification of contrasts?

**Comment by ViralBShah** [comment]
Merge that one right away and then do this?

**Comment by dmbates** [comment]
Yes, I will hold off merging this until the changes for CategoricalArrays etc. have been merged, which I hope will be reasonably soon.

**Comment by dmbates** [comment]
So I am trying to preserve the changes in the `db/modelmatrix` branch and incorporate the changes in the master branch.  I actually went through the many stages of
```
git checkout db/modelmatrix
git rebase master
```
and reconciled the problems.  Now when I try
```
⟫ git status
On branch db/modelmatrix
Your branch and 'origin/db/modelmatrix' have diverged,
and have 3 and 2 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   src/statsmodels/formula.jl
```

I would like to create a pull request from my current copy of the `db/modelmatrix` branch.  Is my best course to save the one file that has been modified, blow away the branch entirely and start over from scratch.

Every time I think I understand how you are supposed to work with git it comes back and bites me.

**Comment by ViralBShah** [comment]
cc @keno or @StefanKarpinski to help ^

**Comment by kleinschmidt** [comment]
If you rebased `db/modelmatrix` onto master, the commits will have different hashes (since they're based in part on the parent). If the contents of the "diverging" commits are the same, you can just `git push --force` to update `origin/db/modelmatrix` with the rebased commits.

**Comment by dmbates** [comment]
Thanks for the suggestions.  In the meantime Andreas created a branch containing the rebasing.  I suspect that the cause of my problems was my local copy of the `db/modelmatrix` being dirty when I started this process.

I hope I have managed to resolve the problem with my usual fallback method of saving a copy of the files that I have changed, removing all my Julia packages and starting over from scratch.  It is surprisingly effective in these cases.
